channels/rail/server/rail_main.c

```diff
@@ -0,0 +1,1616 @@
+
+	if (!s)
+	{
+		WLog_ERR(TAG, "rail_pdu_init failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	rail_write_cloak_order(s, cloak);
+	error = rail_send_pdu(context, s, TS_RAIL_ORDER_CLOAK);
+	Stream_Free(s, TRUE);
+	return error;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rail_send_server_power_display_request(RailServerContext* context,
+        const RAIL_POWER_DISPLAY_REQUEST_ORDER* powerDisplayRequest)
+{
+	wStream* s;
+	UINT error;
+
+	if (!context || !powerDisplayRequest)
+		return ERROR_INVALID_PARAMETER;
+
+	s = rail_pdu_init(RAIL_POWER_DISPLAY_REQUEST_ORDER_LENGTH);
+
+	if (!s)
+	{
+		WLog_ERR(TAG, "rail_pdu_init failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	rail_write_power_display_request_order(s, powerDisplayRequest);
+	error = rail_send_pdu(context, s, TS_RAIL_ORDER_POWER_DISPLAY_REQUEST);
+	Stream_Free(s, TRUE);
+	return error;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rail_send_server_get_app_id_resp(RailServerContext* context,
+                                        const RAIL_GET_APPID_RESP_ORDER* getAppidResp)
+{
+	wStream* s;
+	UINT error;
+
+	if (!context || !getAppidResp)
+		return ERROR_INVALID_PARAMETER;
+
+	s = rail_pdu_init(RAIL_GET_APPID_RESP_ORDER_LENGTH);
+
+	if (!s)
+	{
+		WLog_ERR(TAG, "rail_pdu_init failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	rail_write_get_app_id_resp_order(s, getAppidResp);
+	error = rail_send_pdu(context, s, TS_RAIL_ORDER_GET_APPID_RESP);
+	Stream_Free(s, TRUE);
+	return error;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rail_send_server_get_appid_resp_ex(RailServerContext* context,
+        const RAIL_GET_APPID_RESP_EX_ORDER* getAppidRespEx)
+{
+	wStream* s;
+	UINT error;
+
+	if (!context || !getAppidRespEx)
+		return ERROR_INVALID_PARAMETER;
+
+	s = rail_pdu_init(RAIL_GET_APPID_RESP_EX_ORDER_LENGTH);
+
+	if (!s)
+	{
+		WLog_ERR(TAG, "rail_pdu_init failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	rail_write_get_appid_resp_ex_order(s, getAppidRespEx);
+	error = rail_send_pdu(context, s, TS_RAIL_ORDER_GET_APPID_RESP_EX);
+	Stream_Free(s, TRUE);
+	return error;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rail_read_client_status_order(wStream* s, RAIL_CLIENT_STATUS_ORDER* clientStatus)
+{
+	if (Stream_GetRemainingLength(s) < RAIL_CLIENT_STATUS_ORDER_LENGTH)
+		return ERROR_INVALID_DATA;
+
+	Stream_Read_UINT32(s, clientStatus->flags); /* Flags (4 bytes) */
+	return CHANNEL_RC_OK;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rail_read_exec_order(wStream* s, RAIL_EXEC_ORDER* exec)
+{
+	if (Stream_GetRemainingLength(s) < RAIL_EXEC_ORDER_LENGTH)
+		return ERROR_INVALID_DATA;
+
+	/* TODO: Look into the exec order API change PR before implementing this */
```

Comment:
why u didnt just read the data for now?

Reply:
This is fixed now, pushing soon. I just wanted to address this PDU last (this is variable length too and I didn't wanna start implementing this temporarily and changing later)